include/derecho/core/detail/external_group_impl.hpp

```diff
@@ -0,0 +1,436 @@
+        if(!success) throw derecho_exception("Failed to exchange version hashcodes with the leader! Leader has crashed.");
+        if(leader_version_hashcode != my_version_hashcode) {
+            throw derecho_exception("Unable to connect to Derecho leader because the leader is running on an incompatible platform or used an incompatible compiler.");
+        }
+        success = sock.write(JoinRequest{node_id, true});
+        success = sock.read(leader_response);
+        if(leader_response.code == JoinResponseCode::ID_IN_USE) {
+            dbg_default_error("Error! Leader refused connection because ID {} is already in use!", group.my_id);
+            dbg_default_flush();
+            throw derecho_exception("Leader rejected join, ID already in use.");
+        }
+        success = sock.write(ExternalClientRequest::ESTABLISH_P2P);
+        success = sock.write(getConfUInt16(CONF_DERECHO_EXTERNAL_PORT));
+        if (!success) {
+            throw derecho_exception("Leader crashed before it could send the initial View! Try joining again at the new leader.");
+        }
+
+        assert(dest_node != node_id);
+        sst::add_external_node(dest_node, std::pair<ip_addr_t, uint16_t>{std::get<0>(group.curr_view->member_ips_and_ports[rank]), 
+                                                                std::get<PORT_TYPE::EXTERNAL>(group.curr_view->member_ips_and_ports[rank])});
+        group.p2p_connections->add_connections({dest_node});
+    }
+
+    auto return_pair = wrapped_this->template send<tag>(
+            [this, &dest_node](size_t size) -> char* {
+                const std::size_t max_payload_size = group.max_payload_sizes.at(subgroup_id);
+                if(size <= max_payload_size) {
+                    return (char*)group.get_sendbuffer_ptr(dest_node,
+                                                           sst::REQUEST_TYPE::P2P_REQUEST);
+                } else {
+                    throw derecho_exception("The size of serialized args exceeds the maximum message size.");
+                }
+            },
+            std::forward<Args>(args)...);
+    group.finish_p2p_send(dest_node, subgroup_id, return_pair.pending);
+    return std::move(return_pair.results);
+}
+
+template <typename... ReplicatedTypes>
+ExternalGroup<ReplicatedTypes...>::ExternalGroup(IDeserializationContext* deserialization_context)
+        : my_id(getConfUInt32(CONF_DERECHO_LOCAL_ID)),
+          receivers(new std::decay_t<decltype(*receivers)>()) {
+    if(deserialization_context != nullptr) {
+        rdv.push_back(deserialization_context);
+    }
+#ifdef USE_VERBS_API
+    sst::verbs_initialize(member_ips_and_sst_ports_map,
+                          curr_view->members[curr_view->my_rank]);
+#else
+    sst::lf_initialize({}, my_id, std::map<node_id_t, std::pair<ip_addr_t, uint16_t>>{{my_id, 
+    {getConfString(CONF_DERECHO_LOCAL_IP), getConfUInt16(CONF_DERECHO_EXTERNAL_PORT)}}});
+#endif
+
+    get_view(-1);
```

Comment:
What if this `get_view` fails? (For example, because the current leader rejected this node's ID as already in use). Then `curr_view` will still be null, and you will attempt to deference it on line 87, causing the program to crash with a segfault instead of displaying a useful error message.